### ISSUE TITLE
Questions about protocols

### ISSUE THREAD

**Issue by thomascobb** [bluesky/bluesky#1501]
#1445 implemented some protocols, and I'm in the process of proposing some asyncio compatible versions of them. In doing so I'm trying to understand what is there at present and have some questions:

1. I don't think `Status.add_callback` should never return, should it just return None? https://github.com/bluesky/bluesky/blob/master/bluesky/protocols.py#L14
2. When might `pause` and `resume` never return? https://github.com/bluesky/bluesky/blob/master/bluesky/protocols.py#L195
3. Could we tie down the definition of Configuration down a bit more using a TypedDict? https://github.com/bluesky/bluesky/blob/master/bluesky/protocols.py#L9 Although I was struggling to find where all the optional things that could go in a Configuration were defined...
4. I was thinking about putting `name` and `parent` in a `Named` protocol, and `read_configuration` and `describe_configuration` in a `Configurable` protocol and inheriting from them in `Readable` and `Flyable`, would this work?
5. I was wondering if we needed the inheritance at all, does a `Settable` also have to be `Readable`? Or should the RE do two isinstance checks if it needs something that is both?
6. `runtime_checkable` can't tell the difference between `async def read()` and `def read()`, do we want to require the runengine to check using `inspect.iscoroutinefunction` everything it's going to run or should we use `async def read_async()` instead?
7. I can't actually find anywhere that the RE does `isinstance(Readable)` or similar, did I miss something or is there a plan to use these in the future? 
8. Could we make `trigger` (and other `Status` returning functions) return an `Optional[Status]` rather than a Status that is marked as done immediately?

I've discussed some of these with @tacaswell and @danielballan but was interested to see @ksunden and @untzag 's opinions too

**Comment by ksunden** [comment]
1/2) I think I got confused about `NoReturn` when I was doing this... I now realize that is intended for "always Raises/gets in infinite loop" rather than "has no _explicit_ return", which is what I was trying to indicate, and should just be indicated by `None`

3) I'm open to options, though much of this is quite loosely defined, and I went pretty generic to have _something_ as the return value which current implementations passed.

4) No real argument here, I _almost_ did at least the `Configurable` one at the time, but decided what I had was good enough and could be iterated on.

5) I think the idea is that anything that can be set can also be read, I was mostly going off of prior docs for that being a required relationship.

6) No strong opinion, I think that is more on what makes sense for RE

7) I did not go into changes in the RE when I introduced protocols, I think there is some level of assumption in the RE that anything it is handed is readable (perhaps except flyers)... However, I do believe Queueserver uses isinstance(Readable) to fill out its exposed namespace.

8) I would lean against making Optional[status] because `NullStatus` is pretty easy to use and makes it so that recipients don't need to account for an extra return type.

These are my quick off the cuff thoughts, didn't put too much thought into them, happy to expand if there are followup questions
